CUDADataFormats/TrackingRecHit/interface/TrackingRecHit2DSOAView.h

```diff
@@ -0,0 +1,101 @@
+  __device__ __forceinline__ float& xerrLocal(int i) { return m_xerr[i]; }
+  __device__ __forceinline__ float xerrLocal(int i) const { return __ldg(m_xerr + i); }
+  __device__ __forceinline__ float& yerrLocal(int i) { return m_yerr[i]; }
+  __device__ __forceinline__ float yerrLocal(int i) const { return __ldg(m_yerr + i); }
+
+  __device__ __forceinline__ float& xGlobal(int i) { return m_xg[i]; }
+  __device__ __forceinline__ float xGlobal(int i) const { return __ldg(m_xg + i); }
+  __device__ __forceinline__ float& yGlobal(int i) { return m_yg[i]; }
+  __device__ __forceinline__ float yGlobal(int i) const { return __ldg(m_yg + i); }
+  __device__ __forceinline__ float& zGlobal(int i) { return m_zg[i]; }
+  __device__ __forceinline__ float zGlobal(int i) const { return __ldg(m_zg + i); }
+  __device__ __forceinline__ float& rGlobal(int i) { return m_rg[i]; }
+  __device__ __forceinline__ float rGlobal(int i) const { return __ldg(m_rg + i); }
+
+  __device__ __forceinline__ int16_t& iphi(int i) { return m_iphi[i]; }
+  __device__ __forceinline__ int16_t iphi(int i) const { return __ldg(m_iphi + i); }
+
+  __device__ __forceinline__ int32_t& charge(int i) { return m_charge[i]; }
+  __device__ __forceinline__ int32_t charge(int i) const { return __ldg(m_charge + i); }
+  __device__ __forceinline__ int16_t& clusterSizeX(int i) { return m_xsize[i]; }
+  __device__ __forceinline__ int16_t clusterSizeX(int i) const { return __ldg(m_xsize + i); }
+  __device__ __forceinline__ int16_t& clusterSizeY(int i) { return m_ysize[i]; }
+  __device__ __forceinline__ int16_t clusterSizeY(int i) const { return __ldg(m_ysize + i); }
+  __device__ __forceinline__ uint16_t& detectorIndex(int i) { return m_detInd[i]; }
+  __device__ __forceinline__ uint16_t detectorIndex(int i) const { return __ldg(m_detInd + i); }
+
+  __device__ __forceinline__ pixelCPEforGPU::ParamsOnGPU const& cpeParams() const { return *m_cpeParams; }
+
+  __device__ __forceinline__ uint32_t hitsModuleStart(int i) const { return __ldg(m_hitsModuleStart + i); }
+
+  __device__ __forceinline__ uint32_t* hitsLayerStart() { return m_hitsLayerStart; }
+  __device__ __forceinline__ uint32_t const* hitsLayerStart() const { return m_hitsLayerStart; }
+
+  __device__ __forceinline__ Hist& phiBinner() { return *m_hist; }
+  __device__ __forceinline__ Hist const& phiBinner() const { return *m_hist; }
+
+  __device__ __forceinline__ AverageGeometry& averageGeometry() { return *m_averageGeometry; }
+  __device__ __forceinline__ AverageGeometry const& averageGeometry() const { return *m_averageGeometry; }
+
+private:
+  // local coord
+  float *m_xl, *m_yl;
+  float *m_xerr, *m_yerr;
+
+  // global coord
+  float *m_xg, *m_yg, *m_zg, *m_rg;
+  int16_t* m_iphi;
+
+  // cluster properties
+  int32_t* m_charge;
+  int16_t* m_xsize;
+  int16_t* m_ysize;
+  uint16_t* m_detInd;
+
+  // supporting objects
+  AverageGeometry* m_averageGeometry;  // owned (corrected for beam spot: not sure where to host it otherwise)
```

Comment:
> from the implementation it looks like this is a conditions data, which is apparently recomputed every event.

Correct, only because in principle the beamspot could change between each event. When(/if?) HLT moves to take BeamSpot from the EventSetup, this computation could be done once for the BeamSpot IOV in the EventSetup.

> can this be moved to m_cpeParams (where the uncorrected variant is located)?

No, because `m_cpeParams` comes directly from EventSetup and must not be mutated.

Reply:
> When(/if?) HLT moves to take BeamSpot from the EventSetup, ...

This is planned for Run 3, and the infrastructure part is being tested in the MWGRs.

Reply:
> Correct, only because in principle the beamspot could change between each event. When(/if?) HLT moves to take BeamSpot from the EventSetup, this computation could be done once for the BeamSpot IOV in the EventSetup.

thanks for the reminder that it's event by event in HLT.
A move to the CPE is less directly relevant in this case.
From another comment about likely need to support for another layer of making hits after splitting, the ownership of the m_averageGeometry better still be not with TrackingRecHit2DSOAView

Reply:
Comment updated in https://github.com/cms-patatrack/cmssw/pull/593/commits/782f3bcb52f09ecc2d86ec1bd210a0add4627602